
<!DOCTYPE html>
<html class="">
<head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
  <title>STREAM.FeatureSelection GenePattern Module</title>
  <link href="./assets/docstyle.css" media="all" rel="stylesheet">
  <style type="text/css"></style>
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta content="authenticity_token" name="csrf-param">
</head>
<body >

<script>
</script>
	
	<div class="gp-content-header fluid">
		<div class="container-fluid">
			<div class="row">
				<div class="col-md-12">
			        <h4><img src="./assets/gplogo.png"/>Module Documentation</h4>
                                <h1><big>STREAM.EPGAdjustFinalGraph</big></h1>
				</div>
			</div>
		</div>
	</div>
	<div class="container-fluid">
		<div class="row">
			<div class="col-md-12">
				
<br/>
<dl class="dl-horizontal">
	<dt>LSID</dt>
	<dd>urn:lsid:broad.mit.edu:cancer.software.genepattern.module.analysis:00399</dd>
	<dt>Author(s)</dt>
	<dd>Huidong Chen, Massachussets General Hospital, wrapped as a module by Ted Liefeld, Mesirov Lab, UCSD School of Medicine.</dd>
	<dt>Contact(s)</dt>
	<dd><p>Algorithm and scientific questions:&nbsp;&lt;Huidong.Chen &nbsp;at mgh dot harvard  dot edu&gt;</p>
<p>Module wrapping issues:&nbsp; Ted Liefeld &nbsp;&lt; jliefeld at cloud dot ucsd dot edu&gt;</p>
</dd>
</dl>


<hr>

<h2>Introduction</h2>

<p>STREAM (Single-cell Trajectories Reconstruction, Exploration And Mapping) is an interactive pipeline capable of disentangling and visualizing complex branching trajectories from both single-cell transcriptomic and epigenomic data.  Within GenePattern STREAM is implemented as a collection of modules that cover the entire STREAM processing pipeline to allow individual steps to be performed interactively for data exploration.</p>

<p>STREAM.EPGAdjustFinalGraph facilitates some final adjustments to the graph generated by STREAM.ElasticPrincipalGraph such as optimizing the structure, pruning the final structure or extending leaf nodes top optimize it for single-cell data. </p>

<h2>Algorithm</h2>

<p>
Although ElPiGraph is a general approach to construct principal trees (and other topologies), the obtained structures may not optimally describe biologically relevant trajectories or accurately capture pseudotime information based on single-cell data. Therefore, in addition to the described seeding strategy, in STREAM several single-cell specific optimizations were introduced to the core algorithm of ElPiGraph.
</p>
<p>
<ul><li><b>Control over-branching</b>: A regularization parameter α with range (0,1] was introduced to explicitly control the complexity of the resulting graph structure. Larger values of α lower the propensity of ElPiGraph to introduce branching points. An extreme value close to one prevents the creation of new branches not present in the initial seeding structure. Users can control this parameter based on the expected characteristics of noise and dimensionality of the data. By default, α is set 0.02 and this value was used for all the analyses performed.
</li>
<li><b>Prune branches</b>: The standard elastic principal graph favors harmonicity, i.e., star-shape subgraphs with a central node connected with equally spaced nodes. We have observed that this may lead to trivial branches with few cells. With the pruning grammar rule, STREAM is able to remove branches that are either associated with an excessively small number of cells or that are shorter than a minimal length. This step helps to get rid of spurious and unnecessary branching events that may not reflect real developmental trajectories.
</li>
<li>
<b>Shift branching nodes</b>: To minimize the elastic principal graph energy, ElPiGraph balances the reconstruction error, total length of edges, and the graph harmonicity. However, it may happen that the optimal solution places branching nodes to low-density regions with few cells. This can be hard to interpret biologically since these nodes should correspond to branching cell states within the cell population. With the shift branching grammar rule, each branching node is repositioned to the closest area with higher cell density to better match the most plausible region corresponding to the true branching event.
</li>
<li>
<b>Finetune branching nodes</b>: The standard elastic principal graph procedure well summarizes the principal structure. However, branching sections (i.e., regions close to branching events) may be not described in sufficient details by the obtained curves due to the limited number of nodes used around branching nodes by the global optimization. This grammar rule is able to optimize the space around branching nodes by locally adding a set of nodes in their proximity to better characterize branching events and to improve the pseudotime inference.
</li>
<li>
<b>Extend leaf nodes</b>: The standard elastic principal graph penalizes the total edge length to more robustly capture the main underlying structure. Although this strategy works well when optimizing the graph structure, it may lead to border effects when projecting the data onto the leaf nodes. In fact, edges connecting internal node to leaf nodes rarely extend to the border of their local cloud of points. This is not ideal in the context of pseudotime reconstruction since multiple cells would be mapped to the same leaf node and assigned the same pseudotime. Hence, we extend the principal tree by attaching a new node to each leaf. The location of each node is based on the distribution of points around the corresponding leaf node. This enables the principal graph to better cover terminal cells and infer more accurately their pseudotime.
</li></ul></p>

<h2>References</h2>
<p>H Chen, L Albergante, JY Hsu, CA Lareau, GL Bosco, J Guan, S Zhou, AN Gorban, DE Bauer, MJ Aryee, DM Langenau, A Zinovyev, JD Buenrostro, GC Yuan, L Pinello&nbsp;<a href="https://www.nature.com/articles/s41467-019-09670-4">Single-cell trajectories reconstruction, exploration and mapping of omics data with STREAM. </a> Nature Communications, volume 10, Article number: 1903 (2019)</p>
<p>Nestorowa, S. et al.<a href="https://www.ncbi.nlm.nih.gov/pubmed/27365425"> A single-cell resolution map of mouse hematopoietic stem and progenitor cell differentiation</a>. Blood 128, e20-31 (2016).</p>
<p>Pinello Lab &nbsp; <a href="https://github.com/pinellolab/STREAM">STREAM Github Repository</a></p>

<h2>Input Files</h2>

<ol>
	<li>data file *<br>
 A STREAM pkl file containing an annotated AnnData matrix of gene expression data.</li>
	
</ol>

<h2>Output Files</h2>

<ol>
	<li>&lt;output filename&gt;_stream_result.pkl<br/>
	Output file in STREAM AnnData extended pickle (.pkl) file format suitable for passing to the next step of the STREAM analysis.</li>
    <li>&lt;output filename&gt;_branches.png<bt/>
        Plot of showing the trajectory branches.
        </li>
        <li>&lt;output filename&gt;_branches_with_cells.png<bt/>
        Plot of showing the trajectory branches with the cells positioned on them.
        </li>


</ol>

<h2>Example Data</h2>

<p>Example data can be downloaded from dropbox: <a href="https://www.dropbox.com/sh/dfqht1ob89ku99d/AACI5ZW3aRuq9MhBfSNS_1O_a?dl=0">Stream Example Data</a><br/>
Ref: Nestorowa, S. et al. <a href="https://www.ncbi.nlm.nih.gov/pubmed/27365425">A single-cell resolution map of mouse hematopoietic stem and progenitor cell differentiation</a>. Blood 128, e20-31 (2016).</p>

<h2>Requirements</h2>

<p>GenePattern 3.9.11 or later (dockerized).</p>


<h2>Parameters</h2>

<table class="table table-striped">
	
	<tbody>
		<tr><td colspan=2 ><h4>Inputs and Outputs</h4></td></tr>
		<tr>
			<th>Name</th>
			<th>Description</th>
		</tr>
		<tr>
			<td>data file*</td>
			<td>A STREAM pkl file containing an annotated AnnData matrix of gene expression data/td>
		</tr>
		<tr>
			<td>output filename*</td>
			<td>The output filename prefix.</td>
		</tr>
		<tr><td colspan=2><h4>Elastic Principal Graph</h4>
                </td></tr>
		<tr>
			<th>Name</th>
			<th>Description</th>
		</tr>
                <tr>
                        <td> epg trimming radius*  </td>
                        <td>	Maximal distance from a node to the points it controls in the embedding.   </td>
                </tr>
                <tr>
                        <td>epg alpha*   </td>
                        <td> Alpha parameter of the penalized elastic energy.  </td>
                </tr>
                <tr>
                        <td>epg beta*   </td>
                        <td> Beta parameter of the penalized elastic energy.  </td>
                </tr>
                <tr>
                        <td>epg lambda*   </td>
                        <td>	Lambda parameter used to compute the elastic energy.   </td>
                </tr>
                <tr>
                        <td>epg mu*   </td>
                        <td> Mu parameter used to compute the elastic energy.  </td>
                </tr>

                <tr>
                        <td>epg final energy*   </td>
                        <td>  	Indicate the final elastic energy associated with the configuration. </td>
                </tr>
                <tr><td colspan=2><h4>Optimize Structure</h4>
                </td></tr>
                <tr>
                        <th>Name</th>
                        <th>Description</th>
                </tr>
                <tr>
 			<td>epg max steps*</td>
			<td>The maximum number of iteration steps.</td>
 		</tr>
                <tr>
                        <td>incremental number of nodes*</td>
                        <td>	Incremental number of nodes for elastic principal graph.</td>
                </tr>

              <tr><td colspan=2><h4>Prune Graph</h4>
                </td></tr>
                <tr>
                        <th>Name</th>
                        <th>Description</th>
                </tr>
                <tr>
                        <td>epg collapse mode*</td>
                        <td>The mode used to prune the graph.Choose from {{'PointNumber','PointNumber_Extrema','PointNumber_Leaves','EdgesNumber','EdgesLength'}}. 'PointNumber': branches with less than epg_collapse_par points (points projected on the extreme points are not considered) are removed; 'PointNumber_Extrema', branches with less than epg_collapse_par (points projected on the extreme points are not considered) are removed; 'PointNumber_Leaves', branches with less than epg_collapse_par points (points projected on non-leaf extreme points are not considered) are removed.</td>
                </tr>
                <tr>
                        <td>epg collapse parameter</td>
                        <td>The paramter used to control different modes.</td>
                </tr>

              <tr><td colspan=2><h4>Extend Leaf Nodes</h4>
                </td></tr>
                <tr>
                        <th>Name</th>
                        <th>Description</th>
                </tr>
                <tr>
                        <td>epg extension mode*</td>
                        <td>The mode used to extend the leaves. Choose from {{'QuantDists','QuantCentroid','WeigthedCentroid'}}; 'QuantCentroid':for each leaf node, the extreme points are ordered by their distance from the node and the centroid of the points further than epg_ext_par is returned; 'WeigthedCentroid':for each leaf node, a weight is computed for each points by raising the distance to the epg_ext_par power. Larger epg_ext_par results in a bigger influence of points further than the node; 'QuantDists':for each leaf node, the extreme points are ordered by their distance from the node and the <i>epg extension parameter</i><sup>th</sup> percentile of the points farther than <i>epg extension parameter</i> is returned.</td>
                </tr>
                <tr>
                        <td>epg extension parameter</td>
                        <td>The paramter used to control different modes.</td>
                </tr>
                <tr><td colspan=2><h4>Plotting</h4>
                        Parameters controlling the output figures.
                </td></tr>
		<tr>
                        <th>Name</th>
                        <th>Description</th>
                </tr>
                <tr>
                        <td>num components*   </td>
                        <td>The number of components to be plotted   </td>
                </tr>
                  <tr>
                        <td>component x*   </td>
                        <td>Component used for x-axis in plots   </td>
                </tr>
                  <tr>
                        <td>component y*   </td>
                        <td>Component used for y axis in plots   </td>
                </tr>
  
                <tr>
                        <td>figure height</td>
                        <td>Figure height as used in matplotlib graphs.  Default=8.</td>
                </tr>
                <tr>
                        <td>figure width</td>
                        <td>Figure width as used in matplotlib plots.  Default=8</td>
                </tr>
                <tr>
                        <td>figure legend num columns*   </td>
                        <td>The number of columns that the legend has.   </td>
                </tr>
  

	</tbody>
</table>

<p><span style="color: red;">*</span> - required</p>
			</div>
		</div>
	</div>

	</div>
	<footer class="container fluid gp-footer">
	<div class="fluid text-center">©2006-2019 Regents of the University of California &amp; The Broad Institute of MIT and Harvard</div> 
</footer>



<div id="feedly-mini" title="feedly Mini tookit"></div></body></html>

